File: beltmatic.js
function minOperationsAndPath(target, maxNumber) {
    // Generate numbers from 1 to maxNumber
    const numbers = Array.from({ length: maxNumber }, (_, i) => i + 1);

    // Dictionary to store the minimum operations and path to reach each number
    const operations = {
        0: { ops: 0, path: "", steps: [] }
    };  // It takes 0 operations to reach 0, with an empty path and steps

    // Queue for processing numbers, starting from 0
    const queue = [0];

    while (queue.length > 0) {
        const current = queue.shift();

        // Current number of operations, path, and detailed steps to reach `current`
        const { ops: currentOperations, path: currentPath, steps: currentSteps } = operations[current];

        for (const num of numbers) {
            // Addition
            let newNumber = current + num;
            if (newNumber <= target && !(newNumber in operations)) {
                const newPath = currentPath ? `${currentPath}+${num}` : `${num}`;
                const newSteps = [...currentSteps, `Add ${num} to ${current} to get ${newNumber}`];
                operations[newNumber] = { ops: currentOperations + 1, path: newPath, steps: newSteps };
                queue.push(newNumber);
            }

            // Subtraction
            newNumber = current - num;
            if (newNumber >= 0 && !(newNumber in operations)) {
                const newPath = currentPath ? `${currentPath}-${num}` : `-${num}`;
                const newSteps = [...currentSteps, `Subtract ${num} from ${current} to get ${newNumber}`];
                operations[newNumber] = { ops: currentOperations + 1, path: newPath, steps: newSteps };
                queue.push(newNumber);
            }

            // Multiplication
            newNumber = current * num;
            if (newNumber <= target && !(newNumber in operations)) {
                const newPath = currentPath ? `(${currentPath})*${num}` : `${num}`;
                const newSteps = [...currentSteps, `Multiply ${current} by ${num} to get ${newNumber}`];
                operations[newNumber] = { ops: currentOperations + 1, path: newPath, steps: newSteps };
                queue.push(newNumber);
            }

            // Division
            if (num !== 0 && current % num === 0) {
                newNumber = current / num;
                if (!(newNumber in operations)) {
                    const newPath = currentPath ? `(${currentPath})/${num}` : `1/${num}`;
                    const newSteps = [...currentSteps, `Divide ${current} by ${num} to get ${newNumber}`];
                    operations[newNumber] = { ops: currentOperations + 1, path: newPath, steps: newSteps };
                    queue.push(newNumber);
                }
            }

            // Exponentiation
            if (current !== 0 && num !== 1) {  // Avoid infinite loop with 0^num and 1^num
                try {
                    newNumber = Math.pow(current, num);
                    if (newNumber <= target && !(newNumber in operations)) {
                        const newPath = currentPath ? `(${currentPath})^${num}` : `${current}^${num}`;
                        const newSteps = [...currentSteps, `Raise ${current} to the power of ${num} to get ${newNumber}`];
                        operations[newNumber] = { ops: currentOperations + 1, path: newPath, steps: newSteps };
                        queue.push(newNumber);
                    }
                } catch (e) {
                    continue;  // Skip cases where the exponentiation results in an overflow
                }
            }
        }

        // Early exit if target is reached
        if (target in operations) {
            const { ops, path, steps } = operations[target];
            return { ops, path, steps };
        }
    }

    // If no solution found, return -1 or an indication of impossibility
    if (target in operations) {
        const { ops, path, steps } = operations[target];
        return { ops, path, steps };
    } else {
        return { ops: -1, path: "No solution", steps: [] };
    }
}

document.addEventListener("DOMContentLoaded", () => {
    const maxNumberInput = document.getElementById("max-number");
    const targetInput = document.getElementById("target");
    const resultDiv = document.getElementById("result");

    document.getElementById("calculate").addEventListener("click", () => {
        const maxNumber = parseInt(maxNumberInput.value);
        const target = parseInt(targetInput.value);

        const { ops, path, steps } = minOperationsAndPath(target, maxNumber);
        resultDiv.innerHTML = `
            <p>Minimum operations needed to reach ${target} using numbers from 1 to ${maxNumber}: ${ops}</p>
            <p>Path to reach ${target}: ${path}</p>
            <p>Detailed Steps:</p>
            <ul>${steps.slice(1).map(step => `<li>${step}</li>`).join('')}</ul>
        `;
    });
});

File: beltmatic.py
from collections import deque

# User inputs
max_number = 2
target = 233223


def combine_steps(steps):
    # Combine consecutive additions or subtractions
    combined_steps = []
    i = 0
    while i < len(steps):
        if "Add" in steps[i] or "Subtract" in steps[i]:
            current_op = "Add" if "Add" in steps[i] else "Subtract"
            current_val = int(steps[i].split()[-1])
            combined_step = steps[i]
            i += 1
            while i < len(steps) and current_op in steps[i]:
                current_val += int(steps[i].split()[-1])
                combined_step = f"{current_op} {current_val} to get {current_val}"
                i += 1
            combined_steps.append(combined_step)
        else:
            combined_steps.append(steps[i])
            i += 1
    return combined_steps


def min_operations_and_path(target, max_number):
    numbers = list(range(1, max_number + 1))
    operations = {num: (0, str(num), [f"Start with {num}"]) for num in numbers}
    queue = deque(numbers)
    visited = set(numbers)

    while queue:
        current = queue.popleft()
        current_operations, current_path, current_steps = operations[current]

        for num in numbers:
            # List of potential operations
            potential_operations = [
                (
                    current + num,
                    f"{current_path}+{num}",
                    f"Add {num} to {current} to get {current + num}",
                ),
                (
                    current - num,
                    f"{current_path}-{num}",
                    f"Subtract {num} from {current} to get {current - num}",
                ),
                (
                    current * num,
                    f"({current_path})*{num}",
                    f"Multiply {current} by {num} to get {current * num}",
                ),
            ]

            # Include division only if it is valid
            if num != 0 and current % num == 0:
                potential_operations.append(
                    (
                        current // num,
                        f"({current_path})/{num}",
                        f"Divide {current} by {num} to get {current // num}",
                    )
                )

            # Include exponentiation only if it does not overflow
            if current != 1 and num != 1:
                try:
                    exp_result = current**num
                    if exp_result <= target:
                        potential_operations.append(
                            (
                                exp_result,
                                f"({current_path})^{num}",
                                f"Raise {current} to the power of {num} to get {exp_result}",
                            )
                        )
                except OverflowError:
                    pass

            # Include remainder only if it is valid
            if num != 0:
                remainder = current % num
                potential_operations.append(
                    (
                        remainder,
                        f"({current_path})%{num}",
                        f"Take remainder of {current} divided by {num} to get {remainder}",
                    )
                )

            for new_number, new_path, new_step in potential_operations:
                if new_number <= target and new_number not in visited:
                    new_steps = current_steps + [new_step]
                    operations[new_number] = (
                        current_operations + 1,
                        new_path,
                        combine_steps(new_steps),
                    )
                    queue.append(new_number)
                    visited.add(new_number)

                    if new_number == target:
                        return (
                            current_operations + 1,
                            new_path,
                            combine_steps(new_steps),
                        )

    return -1, "No solution", []


min_ops, path, steps = min_operations_and_path(target, max_number)
print(
    f"Minimum operations needed to reach {target} using numbers from 1 to {max_number}: {min_ops}"
)
print(f"Path to reach {target}: {path}")
print("Detailed Steps:")
for step in steps:
    print(step)
